lib/ai/chat.go

```diff
@@ -0,0 +1,250 @@
+func (chat *Chat) Insert(role string, content string) Message {
+	chat.messages = append(chat.messages, openai.ChatCompletionMessage{
+		Role:    role,
+		Content: content,
+	})
+
+	return Message{
+		Role:    role,
+		Content: content,
+		Idx:     len(chat.messages) - 1,
+	}
+}
+
+// PromptTokens uses the chat's tokenizer to calculate
+// the total number of tokens in the prompt
+//
+// Ref: https://github.com/openai/openai-cookbook/blob/594fc6c952425810e9ea5bd1a275c8ca5f32e8f9/examples/How_to_count_tokens_with_tiktoken.ipynb
+func (chat *Chat) PromptTokens() (int, error) {
+	// perRequest is the number of tokens used up for each completion request
+	const perRequest = 3
+	// perRole is the number of tokens used to encode a message's role
+	const perRole = 1
+	// perMessage is the token "overhead" for each message
+	const perMessage = 3
+
+	sum := perRequest
+	for _, m := range chat.messages {
+		tokens, _, err := chat.tokenizer.Encode(m.Content)
+		if err != nil {
+			return 0, trace.Wrap(err)
+		}
+		sum += len(tokens)
+		sum += perRole
+		sum += perMessage
+	}
+
+	return sum, nil
+}
+
+type Label struct {
+	Key   string `json:"key"`
+	Value string `json:"value"`
+}
+
+type CompletionCommand struct {
+	Command string   `json:"command,omitempty"`
+	Nodes   []string `json:"nodes,omitempty"`
+	Labels  []Label  `json:"labels,omitempty"`
+}
+
+// Summary creates a short summary for the given input.
+func (chat *Chat) Summary(ctx context.Context, message string) (string, error) {
+	resp, err := chat.client.svc.CreateChatCompletion(
+		ctx,
+		openai.ChatCompletionRequest{
+			Model: openai.GPT4,
+			Messages: []openai.ChatCompletionMessage{
+				{Role: openai.ChatMessageRoleSystem, Content: promptSummarizeTitle},
+				{Role: openai.ChatMessageRoleUser, Content: message},
+			},
+		},
+	)
+
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	return resp.Choices[0].Message.Content, nil
+}
+
+// StreamingMessage represents a message that is streamed from the assistant and will later be stored as a normal message in the conversation store.
+type StreamingMessage struct {
+	// Role describes the OpenAI role of the message, i.e it's sender.
+	Role string
+
+	// Idx is a semi-unique ID assigned when loading a conversation so that the UI can group partial messages together.
+	Idx int
+
+	// Chunks is a channel of message chunks that are streamed from the assistant.
+	Chunks <-chan string
+
+	// Error is a channel which may receive one error if the assistant encounters an error while streaming.
+	// Consumers should stop reading from all channels if they receive an error and abort.
+	Error <-chan error
+}
+
+// Complete completes the conversation with a message from the assistant based on the current context.
+// On success, it returns the message and the number of tokens used for the completion.
+func (chat *Chat) Complete(ctx context.Context) (any, int, error) {
```

Comment:
returning an `any` without any documentation of what it can contain makes this a hard API to use. How is the caller meant to use that `any` response? I think at the very least that all possible types returned via this parameter should be documented.

Reply:
Added.